iroh-bytes/src/store/file.rs

```diff
@@ -0,0 +1,2483 @@
+            let Ok(file) = std::fs::File::open(&path) else {
+                return Err(io::Error::new(
+                    io::ErrorKind::NotFound,
+                    format!("file not found: {} size={}", path.display(), outboard_size),
+                )
+                .into());
+            };
+            MemOrFile::File((file, outboard_size))
+        }
+    })
+}
+
+/// Take a possibly incomplete storage and turn it into complete
+fn complete_storage(
+    storage: BaoFileStorage,
+    hash: &Hash,
+    path_options: &PathOptions,
+    inline_options: &InlineOptions,
+) -> ActorResult<std::result::Result<CompleteMemOrFileStorage, CompleteMemOrFileStorage>> {
+    let (data, outboard, _sizes) = match storage {
+        BaoFileStorage::Complete(c) => return Ok(Err(c)),
+        BaoFileStorage::IncompleteMem(storage) => {
+            let (data, outboard, sizes) = storage.into_parts();
+            (
+                MemOrFile::Mem(Bytes::from(data.into_parts().0)),
+                MemOrFile::Mem(Bytes::from(outboard.into_parts().0)),
+                MemOrFile::Mem(Bytes::from(sizes.to_vec()?)),
+            )
+        }
+        BaoFileStorage::IncompleteFile(storage) => {
+            let (data, outboard, sizes) = storage.into_parts();
+            (
+                MemOrFile::File(data),
+                MemOrFile::File(outboard),
+                MemOrFile::File(sizes),
+            )
+        }
+    };
+    let data_size = data.size()?.unwrap();
+    let outboard_size = outboard.size()?.unwrap();
+    // todo: perform more sanity checks if in debug mode
+    debug_assert!(raw_outboard_size(data_size) == outboard_size);
+    // inline data if needed, or write to file if needed
+    let data = if data_size <= inline_options.max_data_inlined {
+        match data {
+            MemOrFile::File(data) => {
+                let mut buf = vec![0; data_size as usize];
+                data.read_at(0, &mut buf)?;
+                let path = path_options.owned_data_path(hash);
+                // this whole file removal thing is not great. It should either fail, or try
+                // again until it works. Maybe have a set of stuff to delete and do it in gc?
+                if let Err(cause) = std::fs::remove_file(path) {
+                    tracing::error!("failed to remove file: {}", cause);
+                };
+                MemOrFile::Mem(Bytes::from(buf))
+            }
+            MemOrFile::Mem(data) => MemOrFile::Mem(data),
+        }
+    } else {
+        match data {
+            MemOrFile::Mem(data) => {
+                let path = path_options.owned_data_path(hash);
+                let file = overwrite_and_sync(&path, &data)?;
+                MemOrFile::File((file, data_size))
+            }
+            MemOrFile::File(data) => MemOrFile::File((data, data_size)),
+        }
+    };
+    // inline outboard if needed, or write to file if needed
+    let outboard = if outboard_size == 0 {
+        Default::default()
```

Comment:
what's the default?

Reply:
An empty outboard. There is a default instance for MemOrFile when the Mem part has a default.